### PR TITLE
Examples/godocbot

### PR DESCRIPTION
**Pull request by droot**
I created this example project from scratch using the `controller-tools scaffold`. I have separated the review in two commits (code and vendor). We can actually drop the `vendor` directory completely and ask users to run `dep ensure` if they want to run the example.

### CODE REVIEW

File: examples/godocbot/pkg/apis/code/v1alpha1/pullrequest_types.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// PullRequestSpec defines the desired state of PullRequest
+type PullRequestSpec struct {
+	// URL of the PULL Request
+	URL string `json:"url"`
+
+	// Latest commit ID on the PR. This is optional.
+	CommitID string `json:"commit_id, omitempty"`
```
Comment (pwittrock) [inline examples/godocbot/pkg/apis/code/v1alpha1/pullrequest_types.go]: nit: delete ws

File: examples/godocbot/pkg/apis/code/v1alpha1/pullrequest_types.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+// PullRequestStatus defines the observed state of PullRequest
+type PullRequestStatus struct {
+	// The URL which is serving the godoc for the PR.
+	GoDocLink string `json:"godoc_link"`
```
Comment (pwittrock) [inline examples/godocbot/pkg/apis/code/v1alpha1/pullrequest_types.go]: json is lower-camel.

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"sigs.k8s.io/controller-tools/examples/godocbot/pkg/apis/code/v1alpha1"
+)
+
+// GithubSyncer implements following functionalities:
+//  - Watches newly created PRs in K8s and updates their commitID by calling
+//  Github
+//  - Periodically updates the PRs in K8s with their commitID in Github.
+type GithubSyncer struct {
+	mgr  manager.Manager
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: Do you need to keep the manager around, or can you just get the client?

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+	ctrl controller.Controller
+	// TODO(droot): take make GithubClient interface to make it easy to test the
+	// syncer
+	ghClient *github.Client
+	// TODO(droot): parameterize the sync duration
+	syncInterval time.Duration
+}
+
+func NewGithubSyncer(mgr manager.Manager, enablePRSync bool, stop <-chan struct{}) (*GithubSyncer, error) {
+	ghClient := github.NewClient(nil)
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: nite: This formatting is a bit hard to read

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+type pullRequestCommitIDReconciler struct {
+	Client client.Client
+	// TODO(droot): take GithubClient interface to improve testability
+	ghClient *github.Client
+}
+
+func (r *pullRequestCommitIDReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.Background()
+
+	log.Printf("reconciling request '%v'", request.NamespacedName)
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: lets delete this debug before we merge

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+	log.Printf("fetching commit id for the PR: %v", prinfo)
+	ghPR, _, err := r.ghClient.PullRequests.Get(context.Background(), prinfo.org, prinfo.repo, int(prinfo.pr))
+	if err != nil {
+		log.Printf("error fetching PR details from github: %v", err)
+		return reconcile.Result{}, err
+	}
+
+	// deep copy ? check if it is still required with pkg/cache or client ?
+	prCopy := pr.DeepCopy()
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: This shouldn't be required

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	if err != nil {
+		log.Printf("error updating PR github: %v", err)
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+// Start periodically syncs PRs in k8s with their commitID in Github.
+func (gs *GithubSyncer) Start(stop <-chan struct{}) {
+	ticker := time.NewTicker(gs.syncInterval)
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: Maybe use:

https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go#L289

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+func (gs *GithubSyncer) syncPullRequests() {
+	prList := &v1alpha1.PullRequestList{}
+	// get pull requests in all namespaces
+	err := gs.mgr.GetClient().List(context.Background(), &client.ListOptions{Namespace: ""}, prList)
+	if err != nil {
+		log.Printf("error fetching all the PRs from k8s: %v", err)
+		return
+	}
+
+	// TODO(droot): simplify the following loop
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: Yes please!

File: examples/godocbot/pkg/apis/code/v1alpha1/pullrequest_types.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (pwittrock) [inline examples/godocbot/pkg/apis/code/v1alpha1/pullrequest_types.go]: Delete this :)

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+			log.Printf("error creating new deployment for PullRequest: %v %v", request, err)
+			return reconcile.Result{}, err
+		}
+		log.Printf("created deployment for %v successfully", request.NamespacedName)
+	}
+
+	prinfo, _ := parsePullRequestURL(pr.Spec.URL)
+	prinfo.commitID = pr.Spec.CommitID
+
+	if len(dp.Spec.Template.Spec.Containers[0].Args) <= 5 || (pr.Spec.CommitID != dp.Spec.Template.Spec.Containers[0].Args[5]) {
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: Maybe there is a better way of doing this.  Patching the args to a cmd is actually pretty common.

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+		if err = r.Client.Update(ctx, dpCopy); err != nil {
+			log.Printf("error updating the deployment for key %s", request.NamespacedName)
+			return reconcile.Result{}, err
+		}
+	}
+
+	if pr.Status.GoDocLink == "" && dp.Status.AvailableReplicas > 0 {
+		log.Printf("deployment became available, updating the godoc link")
+		// update the status
+		prCopy := pr.DeepCopy()
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: this shouldn't be needed

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+		}
+		log.Printf("created deployment for %v successfully", request.NamespacedName)
+	}
+
+	prinfo, _ := parsePullRequestURL(pr.Spec.URL)
+	prinfo.commitID = pr.Spec.CommitID
+
+	if len(dp.Spec.Template.Spec.Containers[0].Args) <= 5 || (pr.Spec.CommitID != dp.Spec.Template.Spec.Containers[0].Args[5]) {
+		// deployment is not updated with latest commit-id
+		dpCopy := dp.DeepCopy()
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: DeepCopy shouldn't be needed if the caching is doing its job

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+	labels := map[string]string{
+		"org":  prinfo.org,
+		"repo": prinfo.repo,
+	}
+
+	tunnelArgs := "80:localhost:6060"
+	tunnelArgs = fmt.Sprintf("%s:%s", prinfo.subdomain(), tunnelArgs)
+
+	dep := &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: This shouldn't be necessary

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+			APIVersion: "apps/v1",
+			Kind:       "Deployment",
+		},
+		ObjectMeta: metav1.ObjectMeta{
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: There should be a library to do the OwnerReference for you now

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+
+// parsePullRequestURL parses given PullRequest URL into prInfo instance.
+// An example pull request URL looks like:
+// https://github.com/kubernetes-sigs/controller-runtime/pull/15
+func parsePullRequestURL(prURL string) (*prInfo, error) {
+	u, err := url.Parse(prURL)
+	if err != nil {
+		return nil, err
+	}
+	parts := strings.Split(strings.TrimPrefix(u.Path, "/"), "/")
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: consider path.Split

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+			return reconcile.Result{}, err
+		}
+	}
+
+	if pr.Status.GoDocLink == "" && dp.Status.AvailableReplicas > 0 {
+		log.Printf("deployment became available, updating the godoc link")
+		// update the status
+		prCopy := pr.DeepCopy()
+		prCopy.Status.GoDocLink = fmt.Sprintf("https://%s.serveo.net/pkg/%s/%s/%s", prinfo.subdomain(), prinfo.host, prinfo.org, prinfo.repo)
+		if err = r.Client.Update(ctx, prCopy); err != nil {
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: Maybe publish an event from an EventRecorder when the link is updated?

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+		log.Printf("created deployment for %v successfully", request.NamespacedName)
+	}
+
+	prinfo, _ := parsePullRequestURL(pr.Spec.URL)
+	prinfo.commitID = pr.Spec.CommitID
+
+	if len(dp.Spec.Template.Spec.Containers[0].Args) <= 5 || (pr.Spec.CommitID != dp.Spec.Template.Spec.Containers[0].Args[5]) {
+		// deployment is not updated with latest commit-id
+		dpCopy := dp.DeepCopy()
+		dpCopy.Spec.Template.Spec.Containers[0].Args = prinfo.godocContainerArgs()
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: Maybe publish an Event fromt he EventRecorder when the commit is updated?

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+	}
+}
+
+// syncPullRequests performs the following:
+//  - fetches all the PRs registered in K8s
+//  - Organize them by orgs/repo so that batch calls can be made
+//  - Fetches details of PRs from Github and updates the commitID of the PRs in
+//  k8s if required.
+// TODO(droot): delete the PRs from K8s if closed in Github.
+func (gs *GithubSyncer) syncPullRequests() {
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: maybe call this `updateAllPullRequestCommits`

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+	prList := &v1alpha1.PullRequestList{}
+	// get pull requests in all namespaces
+	err := gs.mgr.GetClient().List(context.Background(), &client.ListOptions{Namespace: ""}, prList)
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: is there a constant for Namespace All?

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+				continue
+			}
+			for _, ghPR := range ghPRs {
+				ghPRNum := int64(ghPR.GetNumber())
+				if pr, found := prs[ghPRNum]; found {
+					commitID := pr.Spec.CommitID
+					// github PR found in our cluster
+					if ghPR.Head.GetSHA() != commitID {
+						// PR has been updated in GitHub
+						log.Printf("PR Updated: org: %s repo:%s pr: %d commitID: %s ghCommitID: %s \n", org, repo, ghPRNum, commitID, ghPR.Head.GetSHA())
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: All the prints are pretty noisy.  Lets remove them before merging.

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go (hunk 18 above)
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: Should this be in the same object or should it be a different one?  The only thing it seems to share is the client.

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+		log.Printf("Could not find PullRequest %v.\n", request)
+		return reconcile.Result{}, nil
+	}
+
+	if err != nil {
+		log.Printf("Could not fetch PullRequest %v for %+v\n", err, request)
+		return reconcile.Result{}, err
+	}
+
+	if pr.Spec.CommitID != "" {
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: We could use a Predicate to filter these.

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+	// TODO(droot): simplify the following loop
+	orgs := pullRequestByRepoAndOrg(prList)
+	for org, repos := range orgs {
+		for repo, prs := range repos {
+			ghPRs, _, err := gs.ghClient.PullRequests.List(context.Background(), org, repo, nil)
+			if err != nil {
+				log.Printf("cannont get PR list from GH: %v", err)
+				continue
+			}
+			for _, ghPR := range ghPRs {
```
Comment (pwittrock) [inline examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: Consider creation a separate pre-processing function that creates a map[int32]PR

File: examples/godocbot/hack/test.yaml [hunk 23]
```diff
@@ -0,0 +1,6 @@
+apiVersion: code.godocs.io/v1alpha1
```
[the remark this answers is not in the document]
Comment (droot) [reply examples/godocbot/hack/test.yaml]: I am assuming you are referring about the file `test.yaml`. Yes, generating these is super useful.

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go [hunk 24]
```diff
@@ -0,0 +1,265 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"sigs.k8s.io/controller-tools/examples/godocbot/pkg/apis/code/v1alpha1"
+)
+
+// GodocDeployer watches PullRequest object which have a commitID specified in
+// their Spec and deploys a Godoc deployment which runs godoc server for the PR.
+// It watches the PullRequest object for changes in commitID and reconciles the
+// generated godoc deployment.
```
[the remark this answers is not in the document]
Comment (droot) [reply examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: Done.

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go [hunk 25]
```diff
@@ -0,0 +1,265 @@
...
+	}
+
+	if pr.Spec.CommitID == "" {
+		log.Printf("Waiting for PR %s commitID to be updated", request.NamespacedName)
+		return reconcile.Result{}, nil
+	}
+
+	dp := &appsv1.Deployment{}
+	err = r.Client.Get(ctx, request.NamespacedName, dp)
+	if errors.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (droot) [reply examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: Good catch, fixed it.

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go (hunk 12 above)
Comment (droot) [reply examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: Removed.

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go (hunk 11 above)
Comment (droot) [reply examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: Removed.

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go (hunk 13 above)
Comment (droot) [reply examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: hmm... who fills it ? `client` ?

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go (hunk 15 above)
Comment (droot) [reply examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: path.Split gives `dir` and `file` which is different use case.

File: examples/godocbot/pkg/controller/pullrequest/github_syncer.go (hunk 7 above)
Comment (droot) [reply examples/godocbot/pkg/controller/pullrequest/github_syncer.go]: Good pointer, will take a look but keeping it out of this PR for now.

File: examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go (hunk 14 above)
Comment (pwittrock) [reply examples/godocbot/pkg/controller/pullrequest/godoc_deployer.go]: Ping

**Pull request merged**